lib/array.js

```diff
@@ -1,42 +1,76 @@
 'use strict';
 
 const common = require('@metarhia/common');
+const { asyncIter } = require('./async-iterator.js');
+
+const copyClass = name => {
+  switch (name) {
+    case 'Set':
+      return Set;
+    case 'Map':
+      return Map;
+    case 'Array':
+      return Array;
+    case 'String':
+      return String;
+  }
+  return null;
+};
 
-// Asynchronous map (iterate parallel)
-//   items - <Array>, incoming
-//   fn - <Function>, to be executed for each value in the array
-//     current - <any>, current element being processed in the array
-//     callback - <Function>
-//       err - <Error> | <null>
-//       value - <any>
-//   done - <Function>, on done, optional
-//     err - <Error> | <null>
-//     result - <Array>
-const map = (items, fn, done) => {
+const copyInstance = name => {
+  switch (name) {
+    case 'Set':
+      return new Set();
+    case 'Map':
+      return new Map();
+    case 'Array':
+      return [];
+    case 'String':
+      return '';
+  }
+  return null;
+};
+
+const map = (
```

Comment:
```js
// ...
//   items <Array> | <Iterable>,
const map = (items, fn, done) => {
  done = done || common.emptyness;
  const isArray = Array.isArray(items);
  const data = asyncIter(items).map(item => new Promise(
    (resolve, reject) => {
      fn(item, (err, res) => {
        if (err) reject(err);
        else resolve(res);
      });
    })
  );

  const promise = isArray
    ? data.toArray()
    : data.collectTo(items.constructor);

  promise.then(result => done(null, result), error);
};
```

Same for every other function.

Reply:
Should I do the check, if input ```items``` is ```Array```/```Map```/```String```/```Set```?
``` javascript
  if (/* items is not Array/Map/Set/String */) {
    done(null, null);
    return;
  }
```

Reply:
or we just have rule to use only ```Array/String/Map/Set``` and it is enough?